---
title: "Spatial Neighborhood Analysis of Mouse Brain Xenium Data using SpNeigh"
author:
  - name: Jinming Cheng
    affiliation:
    - &wehi_bioinfo Bioinformatics Division, The Walter and Eliza Hall Institute of Medical Research, Parkville, VIC 3052, Australia
    - &uom_med Department of Medical Biology, The University of Melbourne, Parkville, VIC 3010, Australia
    # email: cheng.j@wehi.edu.au
    

date: "`r format(Sys.time(), '%d %B, %Y')`"
vignette: >
  %\VignetteIndexEntry{Spatial Neighborhood Analysis of Mouse Brain Xenium Data using SpNeigh}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}

output:
  BiocStyle::html_document:
    toc_float: false
bibliography-style: plain
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  # cache = TRUE,
  eval = identical(Sys.getenv("IN_PKGDOWN"), "true") ,
  collapse = TRUE,
  comment = "#>",
  out.width='100%',
  dev = "png",
  dpi = 300,
  fig.height = 6*0.7, 
  fig.width = 8*0.7
)
```


# Introduction

The **SpNeigh** package provides a flexible and interpretable framework for analyzing spatial neighborhoods in high-resolution spatial transcriptomics data. Designed with support for technologies such as 10x Xenium, Vizgen MERFISH, and 10x Visium HD, The **SpNeigh** facilitates a range of region-aware and distance-aware analyses. Key features include boundary detection, spatial weight calculation based on proximity to structural landmarks (e.g., region centroid or boundary), spatially-informed differential expression testing using spline-based modeling, and visualization of local cellular interactions.

By integrating spatial geometry with transcriptomic profiles, **SpNeigh** helps uncover biologically meaningful spatial patterns, such as boundary-enriched genes, spatial gradients, and intercellular interactions, enabling deeper insights into tissue organization and cellular communication.


This vignette demonstrates the key functionalities of the **SpNeigh** R package using a real [10x Genomics Xenium Fresh Frozen Mouse Brain Tiny Subset Dataset](https://www.10xgenomics.com/datasets/fresh-frozen-mouse-brain-for-xenium-explorer-demo-1-standard). 

We show how to:

- Identify spatial boundaries and neigbor ring regions
- Explore spatial interaction patterns between clusters
- Perform different expression analysis between two cell populations
- Compute spatial weights from boundary or centroid proximity
- Perform spatial differential expression analysis along spatial gradients
- Peform spatial enrichment analysis along spatial gradients

This workflow highlights the analytical flexibility and biological insights enabled by **SpNeigh**.


# Installation

The **SpNeigh** package can be installed from GitHub by using:

```{r, eval=FALSE}
devtools::install_github("jinming-cheng/SpNeigh")
```


<!-- %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% -->
<!-- %                                                 % -->

# Load packages and data

```{r loadPackage}
library(SpNeigh)
library(ggplot2)
```

Coordinates of mouse brain dataset
```{r}
coords <- readRDS(system.file("extdata", "MouseBrainCoords.rds",
                              package = "SpNeigh"))
head(coords)
```


Ensure the rownames of `coords` are the same with the cell names
```{r}
rownames(coords) <- coords$cell
```


Log normalized expression data generated by `NormalizeData` function in *Seurat* package 
```{r}
logNorm_expr <- readRDS(system.file("extdata", "LogNormExpr.rds", 
                                    package = "SpNeigh"))
class(logNorm_expr)
```


<!-- %                                                 % -->
<!-- %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% -->


<!-- %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% -->
<!-- %                                                 % -->

# Visualize spatial clusters

Plot cells without boundaries
```{r PlotAllCells_NoBoundary, fig.width=8*0.7+1, fig.height=6*0.7}
PlotBoundary(data = coords)
```

Plot cells split by clusters
```{r PlotAllCells_NoBoundary_SplitByCluster, fig.width=8*0.7*1.5, fig.height=6*0.7*1.5}
PlotBoundary(data = coords, split_by = "cluster", angle_x_label = 45)
```


Annotate clusters based on anatomical brain regions
```{r}
coords2 <- coords
new_cell_types = c("0"="Meninges",
                   "1"="Cerebral_cortex",
                   "2"="White_matter",
                   "3"="Cerebral_cortex",
                   "4"="Cerebral_cortex",
                   "5"="Thalamus",
                   "6"="Cerebral_cortex",
                   "7"="Hippocampus",
                   "8"="Hippocampus",
                   "9"="Hippocampus",
                   "10"="White_matter",
                   "11"="Cerebral_cortex")
levels(coords2$cluster) <- new_cell_types
table(coords2$cluster)
```

Plot cells colored by cell types
```{r PlotByBrainStructure, fig.width=8*0.7*1.5, fig.height=6*0.7}
PlotBoundary(data = coords2) + labs(color=NULL)
```

In this vignette, we focus on cluster-level rather than cell typeâ€“level analyses.

<!-- %                                                 % -->
<!-- %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% -->


<!-- %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% -->
<!-- %                                                 % -->

# Neighborhood analysis for cluster 2 cells

## Detect spatial boundaries

Get boundaries of cluster 2. The `eps` and `minPts` parameters can be adjusted to refine the number of regions from detected by `dbscan` (default) subregion clustering method.
```{r}
bon_points_c2 <- GetBoundary(data = coords,
                             one_cluster = 2,
                             eps = 120, 
                             minPts = 10)
table(bon_points_c2$region_id)
```


Add boundaries of cluster 2 to the spatial plot
```{r PlotWithBoundary_C2, fig.width=8*0.7+1, fig.height=6*0.7}
PlotBoundary(coords, boundary = bon_points_c2)
```

Plot boundary regions
```{r PlotRegion_C2,fig.height = 4*0.7,  fig.width = 8*0.7}
bon_polys_c2 <- BuildBoundaryPoly(bon_points_c2)
PlotRegion(bon_polys_c2)
```

## Obtain neighborhood ring regions

Get spatial ring-shaped regions by subtracting the original boundary polygons from their corresponding outer buffered polygons. The `outer_boundary` is automatically computed when it is not supplied.
```{r}
ring_regions <- GetRingRegion(boundary = bon_points_c2)
```


Plot ring regions
```{r PlotRing_C2,fig.height = 4*0.7,  fig.width = 8*0.7}
PlotRegion(ring_regions)
```

## Statistics of cells inside boundaries

Get cells inside the boundaries of cluster 2
```{r}
cells_inside <- GetCellsInside(data = coords,boundary = bon_points_c2)
cells_inside
```

Plot cells inside boundaries
```{r PlotCellsInsideBoundary_C2,fig.height = 4*0.7,  fig.width = 8*0.7}
PlotCellsInside(cells_inside, point_size = 0.1)
```

Obtain statistics of cells inside boundaries of cluster 2
```{r}
stats_cells <- StatsCellsInside(cells_inside)
```


Plot proportion of cells in different clusters for each sub region using bar plot. The `stat_column` is used to plot "proportion" or "count".
```{r PlotStatsBar_Proportion_C2_Boundary}
PlotStatsBar(stats_cells, stat_column = "proportion")
```

Plot proportion of cells in different clusters for each sub region using donut chart. If `plot_donut = FALSE`, pie chart is used.
```{r PlotStats_Donut_C2_Boundary, fig.height = 5*0.7,  fig.width = 8*0.7}
PlotStatsPie(stats_cells, plot_donut = TRUE)
```

## Statistics of cells inside rings

Get cells inside rings for cluster 2
```{r}
cells_ring <- GetCellsInside(data = coords,boundary = ring_regions)
cells_ring
```


Plot cells inside rings
```{r PlotCellsInsideRing_C2,fig.height = 5.5*0.7,  fig.width = 8*0.7}
PlotCellsInside(cells_ring)
```


Obtain statistics of cells inside rings for cluster 2
```{r}
stats_ring <- StatsCellsInside(cells_ring)
```


Plot proportion of cells in different clusters for each sub region using bar plot. 
<!-- The `stat_column` is used to plot "proportion" or "count". -->
```{r PlotStatsBar_Proportion_C2_Ring, fig.height = 6*0.7,  fig.width = 8*0.7 + 1}
PlotStatsBar(stats_ring, stat_column = "proportion")
```



Plot proportion of cells in different clusters for each sub region using donut chart. 
<!-- If `plot_donut = FALSE`, pie chart is used. -->
```{r PlotStats_Donut_C2_Ring, fig.height = 5*0.7,  fig.width = 8*0.7}
PlotStatsPie(stats_ring, plot_donut = TRUE)
```


## Neighborhood interaction of clusters inside boundaries

Compute neighborhood interaction matrix using K-nearest neighbors for cells inside boundaries
```{r}
interaction_matrix <- ComputeSpatialInteractionMatrix(
  subset(coords, cell %in% cells_inside$cell) )
interaction_matrix
```


Heatmap of a row-scaled interaction matrix for cells inside boundaries of cluster 2. The heatmap indicates: (1) cluster 2 is a major neighbor of clusters 2, 3 and 6 for the cells inside the boundaries, and (2) the major neighbor of cluster 2 is cluster 2 itself. 
```{r}
PlotInteractionMatrix(interaction_matrix)
```


## Neighborhood interaction of clusters inside rings

Compute neighborhood interaction matrix using K-nearest neighbors for cells inside rings
```{r}
interaction_matrix <- ComputeSpatialInteractionMatrix(
  subset(coords, cell %in% cells_ring$cell) )
interaction_matrix
```


Heatmap of a row-scaled interaction matrix for cells inside rings for cluster 2. The heatmap reveals that clusters 2, 3, and 4 are spatially adjacent to one another. 
```{r Plot_Heatmap_InteractionMatrix_C2_Ring}
PlotInteractionMatrix(interaction_matrix)
```

The plot of cells inside rings split by clusters further confirms the co-occurrence of clusters 2, 3, 4 in ring region 1.
```{r PlotCellsInside_SplitByCluster_C2_Ring, fig.width=8*0.7*1.5, fig.height=4*0.7*1.5}
PlotCellsInside(cells_ring) + 
  facet_wrap(~cluster) + 
  Seurat::RotatedAxis() + 
  AddBoundaryPoly(ring_regions, linewidth_boundary = 0.1) 
```

<!-- %                                                 % -->
<!-- %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% -->


<!-- %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% -->
<!-- %                                                 % -->

# Differential expression analysis of cluster 2 cells inside and outside boundaries

Most cells in cluster 2 are located in sub region 1. Hence, we focus on the DE analysis between cluster 2 cells inside boundary of sub region 1 and cells in the outside neighboring ring region 1.


Obtain cluster 2 cells inside and outside boundary of sub region
```{r}
cells_in <- subset(cells_inside, region_id==1 & cluster==2)[["cell"]]
cells_out <- subset(cells_ring, region_id==1 & cluster==2)[["cell"]]
```


Perform DE analysis between cells inside boundary and outside boundary using the `limma` framewrok (`lmFit` + `eBayes`)
```{r}
tab <- RunLimmaDE(exp_mat = logNorm_expr,
                  min.pct = 0.25,
                  cells_reference = cells_in,
                  cells_target = cells_out)
```


Top DE genes between cells outside boundary and cells inside boundary. The DE genes are ordered by the abstract value of `logFC` by default.
```{r}
head(tab[, c("gene", "logFC", "adj.P.Val", "pct.reference", "pct.target")])
```


Set a random seed to make PlotExpression results reproducible
```{r}
set.seed(123)
```


Expression of top DE genes in cluster 2 cells outside boundary. The cells are plotted randomly using the given seed when `shuffle = TRUE`. If `shuffle = FALSE`, cells with higher expression values are plotted last (on the top).
```{r Expression_Slc17a7_c2, fig.width = 1.2*(8*0.7 + 1), fig.height = 1.2*(6*0.7)}
PlotExpression(data = coords[colnames(logNorm_expr),], 
               exp_mat = logNorm_expr, 
               genes = tab$gene[1:4], 
               sub_plot = TRUE,
               one_cluster = 2,
               shuffle = TRUE,
               point_size = 0.1,
               angle_x_label = 45)
```


Expression of *Slc17a7* for cluster 2 cells inside and outside boundary of sub region 1. Cluster 2 cells outside the boundary show much higher expression of Slc17a7.
```{r Expression_Slc17a7_InsideOutsideBoundary,fig.height = 4*0.7,  fig.width = 8*0.7}
PlotExpression(data = coords[colnames(logNorm_expr),], 
               exp_mat = logNorm_expr, 
               genes = "Slc17a7", 
               sub_plot = TRUE,
               shuffle = TRUE,
               sub_cells = c(cells_in, cells_out),
               point_size = 0.3) + 
  AddBoundaryPoly(bon_polys_c2[1,], linewidth_boundary = 0.5)
``` 


Expression of a general marker gene *Sox10* for Oligodendrocytes (cluster 2 cells). *Sox10* is expressed in both cluster 2 cells inside and outside the boundary. This suggests that cluster 2 cells located near the boundary may represent intermediate cell states.
```{r PlotExpression_Sox10,fig.height = 4*0.7,  fig.width = 8*0.7}
PlotExpression(data = coords[colnames(logNorm_expr),], 
               exp_mat = logNorm_expr, 
               sub_cells = c(cells_in, cells_out),
               sub_plot = TRUE,
               genes = "Sox10",
               shuffle = TRUE,
               point_size = 0.3) + 
  AddBoundaryPoly(bon_polys_c2[1,], linewidth_boundary = 0.5)
```



<!-- %                                                 % -->
<!-- %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% -->


<!-- %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% -->
<!-- %                                                 % -->

# Spatial differential expression analysis of cells in cluster 0 along spatial weights

## Detect spatial boundaries

Obtain cells in cluster 0
```{r}
cells_c0 <- subset(coords, cluster==0)[,"cell"]
```


Plot cluster 0 cells without boundary
```{r Plot_NoBoundary_C0, fig.width=8*0.7+1, fig.height=6*0.7}
PlotBoundary(coords[cells_c0,])
```


Get boundaries of cluster 0
```{r}
bon_points_c0 <- GetBoundary(data = coords, one_cluster = 0)
```


Add boundaries to the spatial plot of cluster 0 cells
```{r PlotBoundary_C0, fig.width=8*0.7+1, fig.height=6*0.7}
PlotBoundary(data = coords[cells_c0,], 
             boundary = bon_points_c0, 
             linewidth_boundary = 1)
```


## Compute and plot spatial weights

Compute spatial weights based on the distance to boundaries
```{r}
weights_bon <- ComputeBoundaryWeights(data = coords, 
                                      cell_ids = cells_c0,
                                      boundary = bon_points_c0)
```


Alternatively, compute spatial weights based on the distance to centroids
```{r}
weights_cen <- ComputeCentroidWeights(data = coords, cell_ids = cells_c0)
```


Plot boundary weights
```{r, fig.width=8*0.7+1, fig.height=6*0.7}
PlotWeights(data = coords, weights = weights_bon, point_size = 0.8) + 
  labs(title = "Boundary weights")
```


Plot centroid weights
```{r, fig.width=8*0.7+1, fig.height=6*0.7}
PlotWeights(data = coords, weights = weights_cen, point_size = 0.8) + 
  labs(title = "Centroid weights")
```

## Perform spatial differential analysis along boundary weights

Run spatial DE analysis for cluster 0 cells along boundary weights using splines
```{r}
tab_sp <- RunSpatialDE(exp_mat = logNorm_expr[,cells_c0],
                       spatial_distance = weights_bon,
                       cell_ids = cells_c0)
```


Top DE genes along boundary weights. The first spline coefficient (`Z1`) captures the main expression trend along the spatial distance. A positive value indicates increasing expression with distance, while a negative value indicates decreasing expression.
```{r}
head(tab_sp)
```


Expression of top DE genes along boundary weights
```{r Expression_top_genes_boundary, fig.width = 1.2*(8*0.7 + 1), fig.height = 1.2*(6*0.7)}
PlotExpression(data = coords[colnames(logNorm_expr),], 
               exp_mat = logNorm_expr, 
               genes = tab_sp$gene[1:4], 
               sub_plot = TRUE,
               one_cluster = 0,
               shuffle = TRUE,
               point_size = 0.1,
               angle_x_label = 45)
```
Plot scaled average expression along centroid boundary bins for top 10 spatial DE genes. The scaled average expression ranges from 0 to 1 by using min-max normalization method.
```{r Expression_along_boundary_weights_c0, fig.width=8*0.8, fig.height=5*0.8}
PlotSpatialExpression(exp_mat = logNorm_expr[,cells_c0], 
                      spatial_distance = weights_bon,
                      scale_method = "minmax",
                      genes = tab_sp$gene[1:10], 
                      label_x = "Boundary weights")
```


## Perform spatial differential analysis along centroid weights

Run spatial DE analysis for cluster 0 cells along centroid weights using splines
```{r}
tab_sp_cen <- RunSpatialDE(exp_mat = logNorm_expr[,cells_c0],
                           spatial_distance = weights_cen,
                           cell_ids = cells_c0)
```


Top DE genes along centroid weights
```{r}
head(tab_sp_cen)
```


Expression of top DE genes along centroid weights
```{r Expression_top_genes_centroid, fig.width = 1.2*(8*0.7 + 1), fig.height = 1.2*(6*0.7)}
PlotExpression(data = coords[colnames(logNorm_expr),], 
               exp_mat = logNorm_expr, 
               genes = tab_sp_cen$gene[1:4], 
               sub_plot = TRUE,
               one_cluster = 0,
               shuffle = TRUE,
               point_size = 0.1,
               angle_x_label = 45)
```


Plot scaled average expression along centroid weight bins for top 10 spatial DE genes. The scaled average expression ranges from 0 to 1 by using min-max normalization method.
```{r Expression_along_centroid_weights_c0, fig.width=8*0.8, fig.height=5*0.8}
PlotSpatialExpression(exp_mat = logNorm_expr[,cells_c0], 
                      spatial_distance = weights_cen,
                      scale_method = "minmax",
                      genes = tab_sp_cen$gene[1:10], 
                      label_x = "Centroid weights")
```


## Spatial enrichment analysis for each gene

Compute spatial enrichment index (SEI) for each gene based on boundary weights. The SEI reflects the extent to which gene expression is enriched in spatially weighted regions. The result is sorted in descending order by `normalized_SEI`.
```{r}
SEI_bon <- ComputeSpatialEnrichmentIndex(exp_mat = logNorm_expr[,cells_c0], 
                                         weights = weights_bon)
head(SEI_bon)
```

Expression of top genes enriched near boundaries
```{r Expression_top_genes_SEI_boundary, fig.width = 1.2*(8*0.7 + 1), fig.height = 1.2*(6*0.7)}
PlotExpression(data = coords[colnames(logNorm_expr),], 
               exp_mat = logNorm_expr, 
               genes = SEI_bon$gene[1:4], 
               sub_plot = TRUE,
               one_cluster = 0,
               shuffle = TRUE,
               point_size = 0.1,
               angle_x_label = 45)
```


Compute spatial enrichment index (SEI) for each gene based on centroid weights.
```{r}
SEI_cen <- ComputeSpatialEnrichmentIndex(exp_mat = logNorm_expr[,cells_c0], 
                                         weights = weights_cen)
head(SEI_cen)
```

Expression of top genes enriched near the centroid
```{r Expression_top_genes_SEI_centroid, fig.width = 1.2*(8*0.7 + 1), fig.height = 1.2*(6*0.7)}
PlotExpression(data = coords[colnames(logNorm_expr),],
               exp_mat = logNorm_expr, 
               genes = SEI_cen$gene[1:4], 
               sub_plot = TRUE,
               one_cluster = 0,
               shuffle = TRUE,
               point_size = 0.1,
               angle_x_label = 45)
```


<!-- %                                                 % -->
<!-- %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% -->

# Session Info

```{r}
sessionInfo()
```

\pagebreak
